File: desafio-fullcycle/src/domain/customer/entity/customer.ts
import { Address } from "../value-object/address";

export class Customer {
  private _id: string;
  private _name: string;
  private _address!: Address;
  private _active: boolean = false;
  private _rewardPoints: number = 0;

  constructor(id: string, name: string) {
    this._id = id;
    this._name = name;

    this.validate();
  }

  get id(): string {
    return this._id;
  }

  get name(): string {
    return this._name;
  }

  get rewardPoints(): number {
    return this._rewardPoints;
  }

  get Address(): Address {
    return this._address;
  }

  changeAddress(address: Address) {
    this._address = address;
  }

  validate() {
    if (!this._name) {
      throw new Error("Name is required");
    }

    if (!this._id) {
      throw new Error("Id is required");
    }
  }

  changeName(name: string) {
    this._name = name;
    this.validate();
  }

  activate() {
    if (!this._address) {
      throw new Error("Address is required");
    }

    return (this._active = true);
  }

  deactivate() {
    return (this._active = false);
  }

  isActive(): boolean {
    return this._active;
  }

  addRewardPoints(points: number) {
    this._rewardPoints += points;
  }
}

File: entidades-exemplo/src/main/java/com/fredsonchaves07/domain/customer/entity/Customer.java
package com.fredsonchaves07.domain.customer.entity;

public class Customer {

    private String id;

    private String name;

    private Address address;

    private boolean isActive;

    private int rewardsPoints;

    public Customer(String id, String name) {
        this.id = id;
        this.name = name;
        this.isActive = false;
        this.validate();
    }

    private void validate() {
        if (this.name == null || this.name.isBlank()) {
            throw new Error("Name is required");
        }
        if (this.id.isEmpty()) {
            throw new Error("Id is required");
        }
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Address getAddress() {
        return address;
    }

    public Customer setAddress(Address address) {
        this.address = address;
        return this;
    }

    public void activate() {
        this.isActive = true;
    }

    public void addRewardPoints(int points) {
        rewardsPoints += points;
    }

    public int getRewardsPoints() {
        return rewardsPoints;
    }
}
